app/forum.py

```diff
@@ -1,63 +1,80 @@
+from flask import redirect, url_for
 from flask_login import current_user
 from app import db
-from app.models import Group, UserGroup, Post, Thread, User
+from app.models import Group, Post, Thread, User, Quiz
+from app.profile import set_class_code
 from datetime import datetime
 
 def validate_group_link(groupID):
-    group = Group.query.filter_by(id=groupID).first_or_404()
-    if UserGroup.query.filter_by(userID=current_user.id, groupID=groupID).first() is not None:
-        return group
-
-def save_post(form):
-    post = Post(userID=current_user.id, threadID=threadID, 
-                    timestamp=datetime.datetime.now(), 
-                    title=form.title, content=form.post)
+    return Group.query.filter_by(id=groupID).filter(Group.users.any(id=current_user.id)).first_or_404()
+
+def validate_post_link(groupID, threadID, postID):
+    # Check validity of link access first
+    group = validate_group_link(groupID)
+    if group is None:
+        return 
+    thread = Thread.query.filter_by(id=threadID,groupID=groupID).first_or_404()
+    post = Post.query.filter_by(id=postID,threadID=threadID).first_or_404()
+    if current_user.id != post.userID and current_user.urole != 'educator':
+        return 
+    return post
+
+def save_post(form, threadID):
+    post = Post(user=current_user, threadID=threadID, 
+                timestamp=datetime.now(), content=form.post.data)
     db.session.add(post)
     db.session.commit()
 
 
 def create_group(user, name):
     group = Group(name=name)
-    db.session.add(group)
-    db.session.flush()
+    set_class_code(group)
```

Comment:
Why are you setting a new class code in the forum?

Reply:
It's creating a group. Every group needs a code. This was to test out the group in the first place.